### PR TITLE
feat: Objective-C support

### PR DESCRIPTION
**Pull request by falconandy**
### Summary

Adds Objective-C support.

1. Public enums are modified to be int-based instead of string-based.
2. Nullable primitive types are not supported - for Double `Double.nan` is used, for Int - `-1`
3. Custom storages are not supported.
4. Some API overloads are missed in Objective-C to remove error conflicts in Objective-C.
5. Recent changes [fix: encoding Encodable values from [String: Any]](https://github.com/amplitude/Amplitude-Swift/pull/83) don't work with Objective-C types for some reason - right now I reverted the changes to explicit Int32/Int64 switch cases.
6. `disableCarrier` was renamed to `disableTrackCarrier` for consistency with other options - a possible minor breaking change.

Sample code:
```Objective-C
AMPConfiguration* configuration = [[AMPConfiguration alloc] init:@"API-KEY"];
configuration.defaultTracking = AMPDefaultTrackingOptions.ALL;
configuration.logLevel = AMPLogLevelLOG;
configuration.serverZone = AMPServerZoneUS;
configuration.loggerProvider = ^(NSInteger logLevel, NSString* _Nonnull message) {
    NSLog(message);
};
Amplitude* amplitude = [[Amplitude alloc] init:configuration];
[amplitude add:[[AMPPlugin alloc] init:AMPPluginTypeBefore execute:^AMPBaseEvent* _Nullable(AMPBaseEvent* _Nonnull event) {
    [event.eventProperties set:@"test" value:@234];
    event.locationLat = 34;
    event.locationLng = 78;
    return event;
}]];
[amplitude setUserId:@"User-ObjC"];
AMPIdentify* identify = [[AMPIdentify alloc] init];
[identify set:@"prop-1" value:@"value-1"];
[amplitude identify:identify];
[amplitude track:@"Event-A" eventProperties:@{ @"prop1": @"value1" }];
[amplitude track:[[AMPBaseEvent alloc] init:@"Event-B" eventProperties:nil]
    callback:^(AMPBaseEvent* _Nonnull event, NSInteger code, NSString* _Nonnull message) {	    
        NSLog(message);
}];
[amplitude track:[[AMPDeepLinkOpenedEvent alloc] init:@"http://example.com" referrer:@"https://referrer.com"]];
[amplitude flush];
```

### Checklist

* [x] Does your PR title have the correct [title format](https://github.com/amplitude/Amplitude-SDK-Template/blob/main/CONTRIBUTING.md#pr-commit-title-conventions)?
* Does your PR have a breaking change?:  No

### CODE REVIEW

**Review by justin-fiedler — COMMENTED** [review]
@falconandy this looks great! Thank you!

Can you please add a new example app in `Examples/AmplitudeObjCExample` with ObjC implementation and some tests.

**Comment by falconandy** [comment]
@justin-fiedler The test run is fixed

**Comment by falconandy** [comment]
@justin-fiedler @yuhao900914 @Mercy811 
Not sure which option is better (more idiomatic) in case of Objective-C:
For example:
1. Initializer with default name: `[[AmpConfiguration alloc] init:@"API-KEY"]`
2. Initializer with custom name: `[[AmpConfiguration alloc] initWithApiKey:@"API-KEY"]`
3. Class function with custom name: `[AmpConfiguration initWithApiKey:@"API-KEY"]`

WDYT?

**Comment by yuhao900914** [comment]
> @justin-fiedler @yuhao900914 @Mercy811 Not sure which option is better (more idiomatic) in case of Objective-C: For example:
> 
> 1. Initializer with default name: `[[AmpConfiguration alloc] init:@"API-KEY"]`
> 2. Initializer with custom name: `[[AmpConfiguration alloc] initWithApiKey:@"API-KEY"]`
> 3. Class function with custom name: `[AmpConfiguration initWithApiKey:@"API-KEY"]`
> 4. Builders
> 
> WDYT?

I prefer number 3.

**Comment by falconandy** [comment]
@justin-fiedler 
> Can you please do a few minor additions the example before we merge

Done,

File: Examples/AmplitudeObjCExample/AmplitudeObjCExample/AppDelegate.m [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    event.appVersion = @"1.2.3";
+    
+    [amplitude track:event callback:^(AMPBaseEvent* _Nonnull event, NSInteger code, NSString* _Nonnull message) {
+        NSLog(@"%@ - %@", event.eventType, message);
+    }];
+    
+    AMPBaseEvent* deepLinkOpenedEvent = [AMPDeepLinkOpenedEvent initWithUrl:@"http://example.com" referrer:@"https://referrer.com"];
+    [amplitude track:deepLinkOpenedEvent];
+    
+    AMPBaseEvent* screenViewedEvent = [AMPScreenViewedEvent initWithScreenName:@"Settings"];
```
Comment (justin-fiedler) [inline Examples/AmplitudeObjCExample/AmplitudeObjCExample/AppDelegate.m:59]: Not sure why but I am seeing this error:

image

Comment (falconandy) [reply Examples/AmplitudeObjCExample/AmplitudeObjCExample/AppDelegate.m:59]: @justin-fiedler Hope it helps:
image

**Pull request merged**